subprojects/testing-jvm/src/main/java/org/gradle/api/internal/tasks/testing/junit/CategoryFilter.java

```diff
@@ -83,15 +93,29 @@ private boolean shouldRun(final Description description, final Description paren
         return result;
     }
 
-    private boolean matches(final Class<?> category, final Set<Class<?>> categories) {
-        for (Class<?> cls : categories) {
-            if (cls.isAssignableFrom(category)) {
+    private boolean matches(final Class<?> category, final Set<String> categories) {
+        ClassLoader classLoader = category.getClassLoader();
+        for (String cls : categories) {
+            if (loadClass(classLoader, cls).isAssignableFrom(category)) {
```

Comment:
❌ I think you could simply check the class name of the category against the String here. That way you don't need to do any of the classloading magic below.